### ISSUE TITLE
mgr.Terminate() does not wait for .Ack() to mark jobs as processed

### ISSUE THREAD

**Issue by DekzVt** [contribsys/faktory_worker_go#45]
Yesterday I added graceful shutdown to k8s pods that are running faktory_worker_go.

As expected, mgr.Run() waits for all running job processes to complete after receiving the TERM signal and then calls os.Exit(). However, I have noticed that the final job that was completed before the pods exit would stay as 'Busy'. It seems that the os.Exit() is being called before that final ACK message is sent to faktory to mark the job as being processed.

This issue is somewhat related to: https://github.com/contribsys/faktory_worker_go/issues/44 and I do relate to the notion expressed there. It would be nice to run some cleanup code before os.Exit() is called - currently the worker's database dependency does not get closed properly.

Thanks for this wonderful package. I've been using it for months now and have been loving it!

**Comment by mperham** [comment]
Great report, thanks.  I can work on this next week but if you want to send a PR and help out, please do.

**Comment by mperham** [comment]
I've reviewed the code and I don't see how FWG can os.Exit before Ack.  There's a Shutdown Waiter that pauses Exit until all processor goroutines have returned and Ack/Fail is called at the end of those goroutines. Ack is synchronous.

What can happen:

1. FWG starts shutdown but the job doesn't stop
2. The FWG process is killed by k8s after the shutdown timeout.
3. The job is orphaned as Busy until its reservation times out (30 min by default).

The issue is that Go doesn't have a good way (afaik) to kill busy goroutines. I believe the `context` package might be the only clean solution.

**Comment by mperham** [comment]
Thankfully v1.4.0 laid the groundwork for passing a Context to each job.  We can leverage that context.

**Comment by DekzVt** [comment]
Yes i looked through the code and agree that Ack should have been called before os.Exit. If Ack is completely synchronous and doesn't kick off any goroutines that are being orphaned then i'm at a loss to what happened.

I will try to do a bit more debugging. Just to be clear, the job did finish as I was able to confirm that the work was done completely (and we emit a log before the job handler returns that was received). I will need to double confirm that the shutdown timer did not cause the pod to shutdown in between the last job finishing and os.Exit() being called.

If i find an obvious fix in my debugging, i'll definitely submit a PR.

Awesome that passing Context in v1.4.0 will be supported.

**Comment by DekzVt** [comment]
@mperham 

I upgraded to v1.4.0-1 and also hooked up the faktory logger to my logging system so that i could see faktory logs after pods are terminated.

I found that I was getting some 'EOF' and 'broken pipe' error logs from the client library on the final Ack() before a pod was terminated, but my jobs do seem to be making it to the 'processed' page even when those errors are logged.

I will close this issue. Thanks for your help.
